fix: Check property existence for exclude from indexes with wildcard

danieljbruce opened the pull request:
Changes:

Adds a check in three places for the existence of a property in the path provided for exclude from indexes. Also adds test cases to explore what happens when a property is not contained in exclude from indexes, when the property is contained and when it is contained with an array in the path.

Take a look at the changes in the `src` folder. All the changes in the PR involve adding an additional check to see if `entity.properties![firstPathPart]` is defined before accessing a property of `entity.properties![firstPathPart]`. That solves the issue users are seeing where the code throws an error. Instead it just doesn't explore the non-existent property path further and it doesn't add `excludeFromIndexes = true` to any of the entity properties.

The exact test case corresponding to the issue the user reported is at https://github.com/googleapis/nodejs-datastore/pull/1114/files#diff-177d97aec07fd5b22f04049b34d74939b1d8364a7a83e8fb69ee3ee8695105fbR2187-R2198, but similar test cases where the same error would occur were added and source code changes were made to address those test cases.

Fixes #787

danieljbruce committed:
Add a test for testing without mocks

Testing without mocks allows us to look at the code how it is functioning right now so that we can evaluate the intended behavior.

danieljbruce committed:
Do a refactor on the test

Objects shared by both tests should be moved into the outer block so that they can be shared.

danieljbruce committed:
improve the test so that the error bubbles up

Now the test fails properly and passes properly by introducing callbacks

danieljbruce committed:
isolate the latter test right now

The failure should be visible to the user and in the current state it is so now by isolating the test it becomes easier to explore solutions that work.

danieljbruce committed:
test fixes

indent a describe block, modify tests to make sure they pass and refactor a function out that easily allows for testing modified parameters.

danieljbruce committed:
Add two more tests for excludeFromIndexes on array

Add two more tests that address the array case for excludeIndexesFromPath. Also ensure that the conditions which check for these test cases properly ignore exclude from indexes when an array is used

danieljbruce committed:
Move the tests over to the index file

We should not create a nomocks file because we can just group the tests under a unique describe block of test/index.ts.

danieljbruce committed:
Merge branch 'main' into check-property-existence-for-excludeFromIndexes-with-wildcard

ruyadorno commented on src/entity.ts:
```diff
@@ -855,6 +855,7 @@ export namespace entity {
       if (
         firstPathPartIsArray &&
         // check also if the property in question is actually an array value.
+        typeof entity.properties![firstPathPart] !== 'undefined' &&
```
curious: why the check used is `typeof entity.properties![firstPathPart] !== 'undefined'` and not `entity.properties![firstPathPart] !== undefined` ?

danieljbruce replied:
This change causes the test called `should encode a request when using a wildcard` to fail with `TypeError: Cannot read properties of undefined (reading 'arrayValue')`. I guess `typeof entity.properties![firstPathPart]` must evaluate to a string.

danieljbruce replied:
Nevermind, I forgot to remove typeof as per your comment. This change is fine.

danieljbruce replied:
I changed this everywhere and refactored this change out into a variable.

ruyadorno replied:
looks great 👍

ruyadorno commented on src/entity.ts:
```diff
@@ -898,22 +904,26 @@ export namespace entity {
-                  : path + '.*';
-                excludePathFromEntity(parentEntity, newPath);
-              });
+          if (typeof entity.properties![firstPathPart] !== 'undefined') {
+            if (hasWildCard && remainderPath === '*') {
```
I believe this can be simplified to:

```suggestion
          if (typeof entity.properties![firstPathPart] !== 'undefined' && hasWildCard && remainderPath === '*') {
```

danieljbruce replied:
The problem is that the latter if has an else statement.

danieljbruce replied:
I added this change an changed the else to an if else so that the indenting doesn't make it so difficult to read.

danieljbruce committed:
Merge branch 'main' into check-property-existence-for-excludeFromIndexes-with-wildcard

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-datastore into check-property-existence-for-excludeFromIndexes-with-wildcard

danieljbruce committed:
PR update

Remove typeof and remove check for string ‘undefined’

danieljbruce committed:
Merge branch 'check-property-existence-for-excludeFromIndexes-with-wildcard' of https://github.com/danieljbruce/nodejs-datastore into check-property-existence-for-excludeFromIndexes-with-wildcard

danieljbruce committed:
Create a function to refactor the tests

Refactor the tests so that they are more readable. There is a lot of repeated code in each test which makes comparisons hard.

danieljbruce committed:
All tests for this feature now use same script

Replace all tests exercising the feature so that they all use the same test script. This will make the tests way easier to read.

danieljbruce committed:
Add parameterized testing

Add parameterized testing to eliminate repeated code. These tests need to be easier to read for the sake of the reviewer.

danieljbruce committed:
Eliminate function and inline code

This function is only used once now. We should inline this code in the test.

danieljbruce committed:
Add JSON stringify to test title

In the test title include the properties passed in so that it is easy to track what passed/failed.

danieljbruce committed:
Organize code so that similar variables are used

Similar variables should be used together. Organize the code so that there is less confusion about which variables are related.

danieljbruce committed:
inline getExpectedConfig function

The getExpectedConfig does not need to be used anymore because its code can just be inlined in the one place it is used.

danieljbruce committed:
Separate tests into blocks and inline code

Separate the test into blocks and inline some parameters for better test readability.

danieljbruce committed:
Add types to parameters passed into async

Stronger typing makes the code easier to read.

danieljbruce committed:
Eliminate unnecessary variable assignment

Don’t assign to the Datastore variable. This is not necessary. Just use the OriginalDatastore variable directly.

danieljbruce committed:
Ran linter and simplified source changes

The diff will be easier to read if we do not add a nested if statement and instead apply a condition on the if and change the else to else if.

danieljbruce committed:
Add blank line back for easier diff reading

danieljbruce committed:
Try again, eliminate the blank line

danieljbruce committed:
Additional adjustment to entity first path part

Apply the first suggestion in the PR to all examples

danieljbruce committed:
Define isFirstPathPartUndefined

There is a repeated code fragment for checking if the first path part is undefined that we should refactor.

danieljbruce committed:
Rename the variable to align with what it does

danieljbruce committed:
run linter

danieljbruce committed:
Revert "run linter"

This reverts commit a101f27120f780a5561e8dfeea3edbaf0cbdd0d6.

danieljbruce committed:
Revert "Rename the variable to align with what it does"

This reverts commit 4f54d9495b7b37fc0a7a7f5002ec068bf91c7bbc.

danieljbruce committed:
Revert "Define isFirstPathPartUndefined"

This reverts commit df3f3769926f62bf1ed3ea92bb5aef85974a1e42.

danieljbruce committed:
Refactor check out for seeing if defined

We do four checks to see if the first path part is defined. We should refactor these checks out.

danieljbruce committed:
Move comment to more appropriate place

The comment should reflect the next line of code.

ruyadorno commented on test/index.ts:
```diff
@@ -2150,4 +2160,125 @@ describe('Datastore', () => {
+        properties: google.datastore.v1.IValue;
+        entitiesWithoutKey: Entities;
+      }) => {
+        it(`should pass the right properties to upsert on save with parameters: ${JSON.stringify(
+          onSaveTest
+        )}`, async () => {
```
What if we added a `description` with a human-readable, clear description of each scenario (it's probably fine to just reuse the wording for the comments on top of each object definition) instead of just dropping the blob of `JSON.stringify()` which I'm afraid will get really unreadable and harder to maintain with time.

danieljbruce replied:
Sounds good. Then we can use the description as the test name.

danieljbruce replied:
Done

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-datastore into check-property-existence-for-excludeFromIndexes-with-wildcard

# Conflicts:
#	test/index.ts

danieljbruce committed:
Replace comments with description

Replace comments with description property so that the tests will report the description when running the tests instead of the properties in the description.

danieljbruce committed:
Eliminate prefix. Only use description

danieljbruce committed:
fix typo

danieljbruce committed:
lowercase convention

ruyadorno reviewed (approved):
LGTM 👍

danieljbruce committed:
Merge branch 'main' into check-property-existence-for-excludeFromIndexes-with-wildcard
